Throw error when running unsupported features for kernel in FIT format

lucas-akira opened the pull request:
When running certain commands/features, TorizonCore Builder will now first check if the unpacked image has the kernel in FIT format, and abort execution if the current implementation doesn't support it.

This applies to dt, dto, kernel and splash commands, as well as their respective features in the build command.

Related-to: TCB-755

Also, increase the timeout for remote commands sent via SSH to 90 seconds. Some tests in recent pipeline executions were failing early due to the command timeout being reached.

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

lucas-akira committed:
Throw error when running unsupported features for kernel in FIT format

When running certain commands/features, TorizonCore Builder will now
first check if the unpacked image has the kernel in FIT format, and
abort execution if the current implementation doesn't support it.

This applies to dt, dto, kernel and splash commands, as well as their
respective features in the build command.

Related-to: TCB-755

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

lucas-akira committed:
backend/common: Increase timeout for remote commands sent via SSH

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

rborn-tx commented on tcbuilder/backend/common.py:
```diff
@@ -866,3 +874,61 @@ def get_unpacked_uenv_txt_vars(storage_dir, var_list):
+                                text=True, check=True, stdout=subprocess.PIPE)
+
+        # Check if file has the required properties in the root node
+        if not all(prop in output.stdout for prop in required_props):
+            return False
```
Shouldn't this throw an error instead? I mean, at this point we know the file is an FDT, but it could be an actual device-tree blob (DTB), an image-tree blob (ITB) or something unrecognized. The DTB and the unrecognized cases should probably be considered as errors rather than handled as non-FIT by the caller, right?

lucas-akira replied:
Right, it makes sense. Now that you said it, I can't imagine a use case for this function if the binary being checked is in FDT format but isn't a FIT file.

This function wouldn't be used at all if checking a non-FIT FDT file anyway. I'll follow your suggestion.

rborn-tx commented on tcbuilder/cli/dt.py:
```diff
@@ -29,8 +34,14 @@ def do_dt_status(args):
+                                                os.path.join(args.storage_directory, "sysroot",
+                                                             OSTREE_KERNEL_DEPLOY_PATH))
+    if is_file_type_fit(unpacked_kernel_path):
+        raise FeatureNotImplementedError("Command not supported for kernel in FIT image format. "
```
Since FIT = "Flattened **Image** Tree", how about writing "FIT format" rather than "FIT image format" (here and in the other occurrences)?

lucas-akira replied:
Hmm sounds good. I'll change it.

lucas-akira replied:
Done!

rborn-tx commented on tcbuilder/cli/dto.py:
```diff
@@ -50,8 +52,15 @@ def dto_apply(dtos_path, dtb_path, include_dirs, storage_dir,
+        raise InvalidDataError("Device tree overlay customization is not supported for WIC/raw "
+                               "images. Aborting.")
+
+    unpacked_kernel_path = check_if_file_exists(OSTREE_KERNEL_FILENAME,
+                                                os.path.join(storage_dir, "sysroot",
+                                                             OSTREE_KERNEL_DEPLOY_PATH))
```
How about defining a `find_kernel_in_sysroot(storage_dir)` to encapsulate this logic?

Maybe we could even avoid the need for having a `check_if_file_exists()` at all since the globbing performed by that function seems a bit specific to this use case of finding the kernel.

lucas-akira replied:
Right, having the function would be a cleaner implementation.

> Maybe we could even avoid the need for having a check_if_file_exists() at all since the globbing performed by that function seems a bit specific to this use case of finding the kernel.

Thinking now, it's probably better to form the kernel path in sysroot by getting the commit and the kernel version from the OSTree metadata of the unpacked image. I'll use this logic for the new function.

rborn-tx commented on tcbuilder/cli/kernel.py:
```diff
@@ -224,9 +248,17 @@ def do_kernel_clear_custom_args(args):
+        raise InvalidDataError("Kernel customization is not supported for WIC/raw images. "
                                "Aborting.")
 
+
```
Extraneous empty line.

lucas-akira replied:
Thanks, fixed.

rborn-tx reviewed (commented):
@lucas-akira Raise some questions for your consideration.

jsrc27 commented on tcbuilder/backend/common.py:
```diff
@@ -63,7 +63,15 @@
+
+OSTREE_KERNEL_DEPLOY_PATH = "ostree/deploy/torizon/deploy/*/usr/lib/modules/*/"
+
+REMOTE_CMD_TIMEOUT = 90
```
What is the reason for the increased timeout? Was the previous one causing issues by being too short?

lucas-akira replied:
Yeah, some device tests in our pipeline executions were failing due to the command timeout being reached, so I increased it.

jsrc27 commented on tcbuilder/cli/dt.py:
```diff
@@ -29,8 +34,14 @@ def do_dt_status(args):
+                                                os.path.join(args.storage_directory, "sysroot",
+                                                             OSTREE_KERNEL_DEPLOY_PATH))
+    if is_file_type_fit(unpacked_kernel_path):
+        raise FeatureNotImplementedError("Command not supported for kernel in FIT image format. "
```
This is with regards to all the `FeatureNotImplementedError` additions in this PR.

Wouldn't it be more efficent to just have the same generic `Command not supported for kernel in FIT image format. Aborting.` message for everything here.

Instead of having slightly different messages spread across a dozen points in the code.

rborn-tx replied:
Let me chime in to ask: if the message is generic, in case the user is using the "build" command, would it be clear what exactly is not supported?

lucas-akira replied:
I changed the wording according to whether the feature could used in the `build` command as well.: specifically I avoided using the word `command` in this case.

But I can see that it could be easier for maintenance if it had the same generic message for all cases. Maybe it could say 'feature' instead of 'command' like `Feature not supported for kernel in FIT format. Aborting.`

rborn-tx replied:
Let me confirm: if a user tries to set a device-tree through the "build" command, would they be able to understand from the command output that setting the device-tree is not yet supported?

Let's keep in mind that these error messages are just stopgap measures for the short-term, until we implement the features. So, maintenance shouldn't be much of a concern, at least in my view.

lucas-akira replied:
Well, the `build` command has some info logs before entering each subsection that more or less explain what it will do. Assuming the generic error messages are being used, in some cases the `build` logs are clear on what is not supported, like with the splash screen:

```
$ torizoncore-builder build
Building image as per configuration file 'tcbuild.yaml'...

=>> Handling input section
Unpacking Toradex Easy Installer image.
[...]
=>> Handling customization section

=> Setting splash screen
Error: Feature not supported for kernel in FIT format. Aborting.
```

Setting a custom device tree is also clear I'd say:

```
$ torizoncore-builder build
Building image as per configuration file 'tcbuild.yaml'...

=>> Handling input section
Unpacking Toradex Easy Installer image.
[...]
=>> Handling customization section

=> Handling device-tree subsection

=> Selecting custom device-tree 'device-trees/dts-arm32/imx6ull-colibri-eval-v3.dts'
Error: Feature not supported for kernel in FIT format. Aborting.
```

But there are cases where it isn't clear, like when trying to remove an overlay, where it stops at the `device-tree` subsection without explicitly saying that it is about a feature related to overlays:

```
$ torizoncore-builder build
Building image as per configuration file 'tcbuild.yaml'...

=>> Handling input section
Unpacking Toradex Easy Installer image.
[...]
=>> Handling customization section

=> Handling device-tree subsection
Error: Feature not supported for kernel in FIT format. Aborting.
```

rborn-tx replied:
@lucas-akira Thanks for checking. Well, in that case I'd prefer more specific messages as you did originally.

@jsrc27 What do you think?

jsrc27 replied:
I don't have a strong opinion. I just pointed it out cause it looked odd that all the messages were 95% the same.

lucas-akira replied:
I'll keep the messages as is then.

jsrc27 reviewed (commented):
Just left some minor comments.

lucas-akira commented:
FYI I still need to update some tests to consider the case where the kernel in FIT format is unsupported. After that I'll ask for a new round of review @jsrc27 @rborn-tx .

lucas-akira committed:
secboot: Update kernel FIT check

Update the kernel check logic to actually verify if the binary in the
sysroot deployment of the unpacked image follows the FIT format.

This is a more robust method than the previous one that only checked the
value of a U-Boot variable with the kernel image type.

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

lucas-akira committed:
tests/integration: Update tests to check for FIT kernel support

Update some tests to first check if the OS image being tested has
the kernel in FIT format, as some customization features (i.e. dt, dto,
kernel, splash commands) are not currently available in this case. The
tests will be skipped if a FIT kernel is detected.

Also add FIT-only test cases to check if an error message is properly
being shown for these unsupported features.

Related-to: TCB-755

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

lucas-akira committed:
tests: Use a single directory to get both signed/unsigned images

Previously the test workflow looked at a separate directory to get a
signed input image that can be used with the secboot command.

Since the tests can now detect whether the Torizon OS image to be tested
has a FIT kernel or not, we don't need to separate signed and unsigned
images: it is only necessary to check for a FIT kernel and if the image
has the tarball with the signing artifacts.

Signed-off-by: Lucas Akira Morishita <lucas.morishita@toradex.com>

lucas-akira commented:
Alright, I've updated the tests and I've updated the kernel FIT check in `secboot` to use the `is_file_type_fit` function

I've also removed the dependency of the `secboot` tests to a separately provided signed OS image: the test script will now check beforehand if the provided input image has its kernel in FIT format and if it has signing artifacts, and skip or run/test cases accordingly.

rborn-tx commented on tcbuilder/backend/common.py:
```diff
@@ -836,33 +844,94 @@ def check_if_file_exists(filename, directory):
 
-    if not os.path.isfile(uenv_txt_path):
-        raise FileContentMissing("Couldn't find uEnv.txt in unpacked image rootfs! Aborting.")
+    kernel_path = OSTREE_KERNEL_DEPLOY_PATH.replace('*', csum, 1).replace('*', kernel_version, 1)
```
A way to simplify this statement and make it more robust (avoiding depending on the order of the parameters):

```
OSTREE_KERNEL_DEPLOY_PATH = "ostree/deploy/torizon/deploy/{csum}/usr/lib/modules/{kver}/"
kernel_path = OSTREE_KERNEL_DEPLOY_PATH.format(csum=csum, kver=kernel_version)
```

lucas-akira replied:
Nice suggestion. I'll apply it, thanks!

rborn-tx commented on tcbuilder/backend/secboot.py:
```diff
@@ -510,17 +511,14 @@ def check_basic_signing_prerequisites(storage_dir):
-        log.info(f"Detected kernel image format: {uenv_var['kernel_image_type']}")
-    else:
-        log.info("Could not determine the kernel image format")
+    log.info("Checking if the unpacked image has the kernel in FIT format...")
```
Perhaps this could be made a debug message instead?

lucas-akira replied:
Either way would be fine by me, but I have put it as an info log given that a FIT kernel is a strict requirement to run the `secboot` commands, so users would know it always checks this condition beforehand.

jsrc27 replied:
Hmm just to add my view. Right after this an error gets printed if the image is not FIT right? So it's kinda redundant information in that sense.

Like just to imagine:

* Image is FIT they see this log, which can be "ignored".
* Image is not FIT and they see this log and then the error log afterwards reiterating that the image is not FIT.

It's not a big deal to me if this remains as is. But I could see the argument of it not being needed on the default log level.

rborn-tx replied:
Just to clarify I also thought it was a bit redundant, but the reason was the print right after the error check. In the FIT case you'd see:

```
Checking if the unpacked image has the kernel in FIT format...
Found kernel in FIT format.
``` 

while in the non-FIT error case:

```
Checking if the unpacked image has the kernel in FIT format...
Unpacked image does not have the kernel in FIT format. Aborting.
```

lucas-akira replied:
No problem, I'll put it as a debug log message then.

rborn-tx commented on tests/integration/run.sh:
```diff
@@ -228,6 +223,20 @@ rm -rf $SAMPLES_DIR && cp -a ../$SAMPLES_DIR .
+if [ "$IS_WIC" != "1" ] && [ -n "$DEFAULT_TEZI_IMAGE" ]; then
+    torizoncore-builder images --remove-storage unpack "$DEFAULT_TEZI_IMAGE"
+    export IS_DEFAULT_TEZI_IMAGE_FIT="$(set-unpacked-kernel-fit-flag)"
+    if [ "$IS_DEFAULT_TEZI_IMAGE_FIT" = "1" ]; then
```
A more conventional way of doing this would be:

```
export IS_DEFAULT_TEZI_IMAGE_FIT="0"
if unpacked-kernel-in-fit-format; then
    IS_DEFAULT_TEZI_IMAGE_FIT="1"
fi
```

i.e., the exit/return code of the function would indicate the result rather than the text sent to stdout. `unpacked-kernel-in-fit-format` (or whatever you want to call it) could return 0 for FIT, 1 for non-FIT and maybe 2 for invalid FIT (though the condition here would handle an invalid FIT as non-FIT).

lucas-akira replied:
We could handle an invalid FIT as non-FIT, or maybe we could check the status code of the function in case it doesn't return 0, and separate non-FIT cases from any errors like invalid FIT, kernel not found, etc.:

```
export IS_DEFAULT_TEZI_IMAGE_FIT=""
if unpacked-kernel-in-fit-format; then
    IS_DEFAULT_TEZI_IMAGE_FIT="1" # vaild FIT kernel
else
    status=$?
    if [ $status -eq 1 ]; then
        IS_DEFAULT_TEZI_IMAGE_FIT="0" # non-FIT kernel
    else
        # The function would provide the specific error message
        exit $status
    fi
fi
```
What do you think?

rborn-tx replied:
Sounds good to me.

rborn-tx reviewed (commented):
I've added some minor suggestions. Good improvements from the previous version, btw!

rborn-tx commented on tests/integration/run.sh:
```diff
@@ -223,11 +218,34 @@ fi
+    torizoncore-builder images --remove-storage unpack "$DEFAULT_TEZI_IMAGE"
+    if unpacked-kernel-in-fit-format; then
+        echo "Image has kernel in FIT format"
+        IS_DEFAULT_TEZI_IMAGE_FIT="1" # vaild FIT kernel
```
vaild -> valid

lucas-akira replied:
Fixed, thanks!

rborn-tx reviewed (commented):
Found a typo
